<a href="https://colab.research.google.com/github/mayankbrn/9.7_Delhivery_feature_engineering/blob/MA_working/07_Delhivery_Feature_Engineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Business Case: Delhivery - Feature Engineering



## About Delhivery

Delhivery is the largest and fastest-growing fully integrated player in India by revenue in Fiscal 2021. They aim to build the operating system for commerce, through a combination of world-class infrastructure, logistics operations of the highest quality, and cutting-edge engineering and technology capabilities.

The Data team builds intelligence and capabilities using this data that helps them to widen the gap between the quality, efficiency, and profitability of their business versus their competitors.

### How can you help here?



The company wants to understand and process the data coming out of data engineering pipelines:

• Clean, sanitize and manipulate data to get useful features out of raw fields

• Make sense out of the raw data and help the data science team to build forecasting models on it

### Dataset

Dataset Link: [Delhivery data](https://drive.google.com/file/d/1ZkF2gGCDkjwQgOTGVBpsqhPpSGg1Fybb/view?usp=drive_link)



### Column Profiling

The dataset at the heart of this exploration. Here are some of the key features:

- **data**: tells whether the data is testing or training data.
- **trip_creation_time**: Timestamp of trip creation.
- **route_schedule_uuid**: Unique Id for a particular route schedule.
- **route_type**: Transportation type.
- **FTL**: Full Truck Load - FTL shipments get to the destination sooner, as the truck is making no other pickups or drop-offs along the way.
- **Carting**: Handling system consisting of small vehicles (carts).
- **trip_uuid**: Unique ID given to a particular trip (A trip may include different source and destination centers).
- **source_center**: Source ID of trip origin.
- **source_name**: Source Name of trip origin.
- **destination_center**: Destination ID.
- **destination_name**: Destination Name.
- **od_start_time**: Trip start time.
- **od_end_time**: Trip end time.
- **start_scan_to_end_scan**: Time taken to deliver from source to destination.
- **is_cutoff**: Unknown field.
- **cutoff_factor**: Unknown field.
- **cutoff_timestamp**: Unknown field.
- **actual_distance_to_destination**: Distance in Kms between source and destination warehouse.
- **actual_time**: Actual time taken to complete the delivery (Cumulative).
- **osrm_time**: An open-source routing engine time calculator which computes the shortest path between points in a given map (Includes usual traffic, distance through major and minor roads) and gives the time (Cumulative).
- **osrm_distance**: An open-source routing engine which computes the shortest path between points in a given map (Includes usual traffic, distance through major and minor roads) (Cumulative).
- **factor**: Unknown field.
- **segment_actual_time**: This is a segment time. Time taken by the subset of the package delivery.
- **segment_osrm_time**: This is the OSRM segment time. Time taken by the subset of the package delivery.
- **segment_osrm_distance**: This is the OSRM distance. Distance covered by the subset of the package delivery.
- **segment_factor**: Unknown field.

### Concepts Used

- Feature Creation
- Relationship between Features
- Column Normalization /Column Standardization
- Handling categorical values
- Missing values - Outlier treatment / Types of outliers

### How to begin:

Since delivery details of one package are divided into several rows (think of it as connecting flights to reach a particular destination). Now think about how we should treat their fields if we combine these rows? What aggregation would make sense if we merge. What would happen to the numeric fields if we merge the rows?

#### Hint:


You can use inbuilt functions like `groupby` and aggregations like `sum()`, `cumsum()` to merge some rows based on their
1. `Trip_uuid`, `Source ID` and `Destination ID`
2. Further aggregate on the basis of just `Trip_uuid`. You can also keep the first and last values for some numeric/categorical fields if aggregating them won’t make sense.

#### Basic data cleaning and exploration:

- Handle missing values in the data.
- Analyze the structure of the data.
- Try merging the rows using the hint mentioned above.
- Build some features to prepare the data for actual analysis. Extract features from the below fields:
  - **Destination Name**: Split and extract features out of destination. City-place-code (State)
  - **Source Name**: Split and extract features out of destination. City-place-code (State)
  - **Trip_creation_time**: Extract features like month, year and day etc

#### In-depth analysis and feature engineering:

- Calculate the time taken between `od_start_time` and `od_end_time` and keep it as a feature. Drop the original columns, if required
- Compare the difference between Point a. and `start_scan_to_end_scan`. Do hypothesis testing/ Visual analysis to check.
- Do hypothesis testing/ visual analysis between `actual_time` aggregated value and `OSRM time` aggregated value (aggregated values are the values you’ll get after merging the rows on the basis of `trip_uuid`)
- Do hypothesis testing/ visual analysis between `actual_time` aggregated value and `segment actual time` aggregated value (aggregated values are the values you’ll get after merging the rows on the basis of `trip_uuid`)
- Do hypothesis testing/ visual analysis between `osrm distance` aggregated value and `segment osrm distance` aggregated value (aggregated values are the values you’ll get after merging the rows on the basis of `trip_uuid`)
- Do hypothesis testing/ visual analysis between `osrm time` aggregated value and `segment osrm time` aggregated value (aggregated values are the values you’ll get after merging the rows on the basis of `trip_uuid`)
- Find outliers in the numerical variables (you might find outliers in almost all the variables), and check it using visual analysis
- Handle the outliers using the **IQR method**.
- Do one-hot encoding of categorical variables (like `route_type`)
- Normalize/ Standardize the numerical features using `MinMaxScaler` or `StandardScaler`.

### Evaluation Criteria (100 Points):


- Define Problem Statement and Perform Exploratory Data Analysis (10 points)

  - **Definition of Problem** (as per given problem statement with additional views)
  - **Observations** on:
    - Shape of data
    - Data types of all the attributes
    - Conversion of categorical attributes to 'category' (if required)
    - Missing value detection
    - Statistical summary
  - **Visual Analysis**:
    - Distribution plots of all the continuous variable(s)
    - Boxplots of all the categorical variables
  - **Insights** based on EDA
  - **Comments** on:
    - Range of attributes
    - Outliers of various attributes
    - Distribution of the variables and relationship between them
  - **Comments** for each univariate and bivariate plot

- Feature Creation (10 Points)
- Merging of Rows and Aggregation of Fields (10 Points)
- Comparison & Visualization of Time and Distance Fields (10 Points)
- Missing Values Treatment & Outlier Treatment (10 Points)
- Checking Relationship Between Aggregated Fields (10 Points)
- Handling Categorical Values (10 Points)
- Column Normalization / Column Standardization (10 Points)

- Business Insights (10 Points):
  Should include patterns observed in the data along with what you can infer from it. Examples:
  - Check from where most orders are coming from (State, Corridor, etc.)
  - Busiest corridor, average distance between them, average time taken

- Recommendations (10 Points)
  Actionable items for business. No technical jargon. No complications. Simple action items that everyone can understand.

## Solution

### Basic data cleaning and exploration:

In [323]:
#importing the relevent libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [324]:
#importing the csv file from google drive
!gdown 1ZkF2gGCDkjwQgOTGVBpsqhPpSGg1Fybb

Downloading...
From: https://drive.google.com/uc?id=1ZkF2gGCDkjwQgOTGVBpsqhPpSGg1Fybb
To: /content/07_delhivery_data.csv
100% 55.6M/55.6M [00:00<00:00, 89.8MB/s]


In [325]:
#load the csv into dataframe
df = pd.read_csv('07_delhivery_data.csv')

In [326]:
df.head(5)

data          trip_creation_time  \
0  training  2018-09-20 02:35:36.476840   
1  training  2018-09-20 02:35:36.476840   
2  training  2018-09-20 02:35:36.476840   
3  training  2018-09-20 02:35:36.476840   
4  training  2018-09-20 02:35:36.476840   

                                 route_schedule_uuid route_type  \
0  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
1  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
2  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
3  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
4  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   

                 trip_uuid source_center                 source_name  \
0  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
1  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
2  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
3  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
4  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   

  destination_center               destination_name  \
0       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
1       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
2       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
3       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
4       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   

                od_start_time  ...            cutoff_timestamp  \
0  2018-09-20 03:21:32.418600  ...         2018-09-20 04:27:55   
1  2018-09-20 03:21:32.418600  ...         2018-09-20 04:17:55   
2  2018-09-20 03:21:32.418600  ...  2018-09-20 04:01:19.505586   
3  2018-09-20 03:21:32.418600  ...         2018-09-20 03:39:57   
4  2018-09-20 03:21:32.418600  ...         2018-09-20 03:33:55   

   actual_distance_to_destination  actual_time  osrm_time osrm_distance  \
0                       10.435660         14.0       11.0       11.9653   
1                       18.936842         24.0       20.0       21.7243   
2                       27.637279         40.0       28.0       32.5395   
3                       36.118028         62.0       40.0       45.5620   
4                       39.386040         68.0       44.0       54.2181   

     factor  segment_actual_time  segment_osrm_time  segment_osrm_distance  \
0  1.272727                 14.0               11.0                11.9653   
1  1.200000                 10.0                9.0                 9.7590   
2  1.428571                 16.0                7.0                10.8152   
3  1.550000                 21.0               12.0                13.0224   
4  1.545455                  6.0                5.0                 3.9153   

   segment_factor  
0        1.272727  
1        1.111111  
2        2.285714  
3        1.750000  
4        1.200000  

[5 rows x 24 columns]

#### Dropping the unknown columns

In [327]:
# all the columns which are marked as unknown in the column profiling we can remove them

unknown_fields = ['is_cutoff', 'cutoff_factor', 'cutoff_timestamp', 'factor', 'segment_factor']
df.drop(unknown_fields, axis = 1, inplace = True)

We can see that time is considered as object and float in no of columns

In [328]:
df.head(5)

data          trip_creation_time  \
0  training  2018-09-20 02:35:36.476840   
1  training  2018-09-20 02:35:36.476840   
2  training  2018-09-20 02:35:36.476840   
3  training  2018-09-20 02:35:36.476840   
4  training  2018-09-20 02:35:36.476840   

                                 route_schedule_uuid route_type  \
0  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
1  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
2  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
3  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
4  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   

                 trip_uuid source_center                 source_name  \
0  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
1  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
2  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
3  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
4  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   

  destination_center               destination_name  \
0       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
1       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
2       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
3       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
4       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   

                od_start_time                 od_end_time  \
0  2018-09-20 03:21:32.418600  2018-09-20 04:47:45.236797   
1  2018-09-20 03:21:32.418600  2018-09-20 04:47:45.236797   
2  2018-09-20 03:21:32.418600  2018-09-20 04:47:45.236797   
3  2018-09-20 03:21:32.418600  2018-09-20 04:47:45.236797   
4  2018-09-20 03:21:32.418600  2018-09-20 04:47:45.236797   

   start_scan_to_end_scan  actual_distance_to_destination  actual_time  \
0                    86.0                       10.435660         14.0   
1                    86.0                       18.936842         24.0   
2                    86.0                       27.637279         40.0   
3                    86.0                       36.118028         62.0   
4                    86.0                       39.386040         68.0   

   osrm_time  osrm_distance  segment_actual_time  segment_osrm_time  \
0       11.0        11.9653                 14.0               11.0   
1       20.0        21.7243                 10.0                9.0   
2       28.0        32.5395                 16.0                7.0   
3       40.0        45.5620                 21.0               12.0   
4       44.0        54.2181                  6.0                5.0   

   segment_osrm_distance  
0                11.9653  
1                 9.7590  
2                10.8152  
3                13.0224  
4                 3.9153

#### Analyzing the structure of data

In [329]:
#checking Null values in the columns
df.isna().sum()

data                                0
trip_creation_time                  0
route_schedule_uuid                 0
route_type                          0
trip_uuid                           0
source_center                       0
source_name                       293
destination_center                  0
destination_name                  261
od_start_time                       0
od_end_time                         0
start_scan_to_end_scan              0
actual_distance_to_destination      0
actual_time                         0
osrm_time                           0
osrm_distance                       0
segment_actual_time                 0
segment_osrm_time                   0
segment_osrm_distance               0
dtype: int64

'Source name' and 'Destination name' have some missing values.

In [330]:
#unique values in the columns
df.nunique()

data                                   2
trip_creation_time                 14817
route_schedule_uuid                 1504
route_type                             2
trip_uuid                          14817
source_center                       1508
source_name                         1498
destination_center                  1481
destination_name                    1468
od_start_time                      26369
od_end_time                        26369
start_scan_to_end_scan              1915
actual_distance_to_destination    144515
actual_time                         3182
osrm_time                           1531
osrm_distance                     138046
segment_actual_time                  747
segment_osrm_time                    214
segment_osrm_distance             113799
dtype: int64

In [331]:
# statistical summary of data - Numercial columns
df.describe()

start_scan_to_end_scan  actual_distance_to_destination    actual_time  \
count           144867.000000                   144867.000000  144867.000000   
mean               961.262986                      234.073372     416.927527   
std               1037.012769                      344.990009     598.103621   
min                 20.000000                        9.000045       9.000000   
25%                161.000000                       23.355874      51.000000   
50%                449.000000                       66.126571     132.000000   
75%               1634.000000                      286.708875     513.000000   
max               7898.000000                     1927.447705    4532.000000   

           osrm_time  osrm_distance  segment_actual_time  segment_osrm_time  \
count  144867.000000  144867.000000        144867.000000      144867.000000   
mean      213.868272     284.771297            36.196111          18.507548   
std       308.011085     421.119294            53.571158          14.775960   
min         6.000000       9.008200          -244.000000           0.000000   
25%        27.000000      29.914700            20.000000          11.000000   
50%        64.000000      78.525800            29.000000          17.000000   
75%       257.000000     343.193250            40.000000          22.000000   
max      1686.000000    2326.199100          3051.000000        1611.000000   

       segment_osrm_distance  
count           144867.00000  
mean                22.82902  
std                 17.86066  
min                  0.00000  
25%                 12.07010  
50%                 23.51300  
75%                 27.81325  
max               2191.40370

In [332]:
#statistical summary of data - categorical columns
df.describe(include = object)

data          trip_creation_time  \
count     144867                      144867   
unique         2                       14817   
top     training  2018-09-28 05:23:15.359220   
freq      104858                         101   

                                      route_schedule_uuid route_type  \
count                                              144867     144867   
unique                                               1504          2   
top     thanos::sroute:4029a8a2-6c74-4b7e-a6d8-f9e069f...        FTL   
freq                                                 1812      99660   

                      trip_uuid source_center                    source_name  \
count                    144867        144867                         144574   
unique                    14817          1508                           1498   
top     trip-153811219535896559  IND000000ACB  Gurgaon_Bilaspur_HB (Haryana)   
freq                        101         23347                          23347   

       destination_center               destination_name  \
count              144867                         144606   
unique               1481                           1468   
top          IND000000ACB  Gurgaon_Bilaspur_HB (Haryana)   
freq                15192                          15192   

                     od_start_time                 od_end_time  
count                       144867                      144867  
unique                       26369                       26369  
top     2018-09-21 18:37:09.322207  2018-09-24 09:59:15.691618  
freq                            81                          81

In [333]:
#shape of data
df.shape

(144867, 19)

There are 19 columns and 144867 rows in the dataset.

In [334]:
#checking the datatype of columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144867 entries, 0 to 144866
Data columns (total 19 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   data                            144867 non-null  object 
 1   trip_creation_time              144867 non-null  object 
 2   route_schedule_uuid             144867 non-null  object 
 3   route_type                      144867 non-null  object 
 4   trip_uuid                       144867 non-null  object 
 5   source_center                   144867 non-null  object 
 6   source_name                     144574 non-null  object 
 7   destination_center              144867 non-null  object 
 8   destination_name                144606 non-null  object 
 9   od_start_time                   144867 non-null  object 
 10  od_end_time                     144867 non-null  object 
 11  start_scan_to_end_scan          144867 non-null  float64
 12  actual_distance_

#### Changing he datatype of columns

In [335]:
#converting the data and route_type as categorical columns

df["data"] = df["data"].astype("category")
df["route_type"] = df["route_type"].astype("category")

In [336]:
# There are few columns which represent time but they in object datatype format those can be convereted to datetime column

datetype_column = ['trip_creation_time', 'od_start_time', 'od_end_time']

for i in datetype_column:
  df[i] = pd.to_datetime(df[i]) #passing the column one by one in the for loop

In [337]:
#checking the column datatypes again
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144867 entries, 0 to 144866
Data columns (total 19 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   data                            144867 non-null  category      
 1   trip_creation_time              144867 non-null  datetime64[ns]
 2   route_schedule_uuid             144867 non-null  object        
 3   route_type                      144867 non-null  category      
 4   trip_uuid                       144867 non-null  object        
 5   source_center                   144867 non-null  object        
 6   source_name                     144574 non-null  object        
 7   destination_center              144867 non-null  object        
 8   destination_name                144606 non-null  object        
 9   od_start_time                   144867 non-null  datetime64[ns]
 10  od_end_time                     144867 non-null  datetim

#### Handling missing values

In [338]:
# checking for the source_center  for which source_name is null

center_for_missing_source_name = df[df['source_name'].isna()]['source_center'].unique()
center_for_missing_source_name

array(['IND342902A1B', 'IND577116AAA', 'IND282002AAD', 'IND465333A1B',
       'IND841301AAC', 'IND509103AAC', 'IND126116AAA', 'IND331022A1B',
       'IND505326AAB', 'IND852118A1B'], dtype=object)

This give us all the source_center names for which source_name is not available.
Let's check if these source_center name is having source_name

In [339]:
# Checking if we can get the source_name for above source_centers from other rows in the data
# In short, it's selecting rows where "source_name" has a value, but the "source_center" is associated with some other rows where "source_name" is missing (null).

df[(df['source_name'].notnull()) & (df['source_center'].isin(df[df['source_name'].isnull()]))]

Empty DataFrame
Columns: [data, trip_creation_time, route_schedule_uuid, route_type, trip_uuid, source_center, source_name, destination_center, destination_name, od_start_time, od_end_time, start_scan_to_end_scan, actual_distance_to_destination, actual_time, osrm_time, osrm_distance, segment_actual_time, segment_osrm_time, segment_osrm_distance]
Index: []

It means that source_name for missing values is not available in other rows as well.

In [340]:
#checking for destination_center for which destination_name is null
center_for_missing_destination_name = df[df['destination_name'].isna()]['destination_center'].unique()
center_for_missing_destination_name

array(['IND342902A1B', 'IND577116AAA', 'IND282002AAD', 'IND465333A1B',
       'IND841301AAC', 'IND505326AAB', 'IND852118A1B', 'IND126116AAA',
       'IND509103AAC', 'IND221005A1A', 'IND250002AAC', 'IND331001A1C',
       'IND122015AAC'], dtype=object)

In [341]:
# Checking if we can get the destintion_name for above destinatio_centers from other rows in the data
# In short, it's selecting rows where "destination_name" has a value, but the "destination_center" is associated with some other rows where "destination_name" is missing (null).

df[(df['destination_name'].notnull()) & (df['destination_center'].isin(df[df['destination_name'].isnull()]))]

Empty DataFrame
Columns: [data, trip_creation_time, route_schedule_uuid, route_type, trip_uuid, source_center, source_name, destination_center, destination_name, od_start_time, od_end_time, start_scan_to_end_scan, actual_distance_to_destination, actual_time, osrm_time, osrm_distance, segment_actual_time, segment_osrm_time, segment_osrm_distance]
Index: []

#### Removing incorrect data

It means that destination_name for missing values is not available in other rows as well.

In [342]:
# We can see from the describe that the values of segment actual time is negative, which can't be true so lets drop that data
df.drop(df[df['segment_actual_time']<0].index, inplace = True)

In [343]:
#now checking the describe data
df.describe()

trip_creation_time                  od_start_time  \
count                         144846                         144846   
mean   2018-09-22 13:34:27.259366400  2018-09-22 18:02:50.434589952   
min       2018-09-12 00:00:16.535741     2018-09-12 00:00:16.535741   
25%    2018-09-17 03:20:51.775845888  2018-09-17 08:05:40.886155008   
50%    2018-09-22 04:24:27.932764928  2018-09-22 08:52:50.639791104   
75%    2018-09-27 17:57:56.350054912  2018-09-27 22:41:50.285857024   
max       2018-10-03 23:59:42.701692     2018-10-06 04:27:23.392375   
std                              NaN                            NaN   

                         od_end_time  start_scan_to_end_scan  \
count                         144846           144846.000000   
mean   2018-09-23 10:04:33.787580160              961.226537   
min       2018-09-12 00:50:10.814399               20.000000   
25%    2018-09-18 01:48:06.410121984              161.000000   
50%    2018-09-23 03:13:03.520212992              449.000000   
75%    2018-09-28 12:49:06.054018048             1634.000000   
max       2018-10-08 03:00:24.353479             7898.000000   
std                              NaN             1036.993595   

       actual_distance_to_destination    actual_time      osrm_time  \
count                   144846.000000  144846.000000  144846.000000   
mean                       234.057171     416.908724     213.853002   
min                          9.000045       9.000000       6.000000   
25%                         23.354927      51.000000      27.000000   
50%                         66.126234     132.000000      64.000000   
75%                        286.706673     513.000000     257.000000   
max                       1927.447705    4532.000000    1686.000000   
std                        344.974984     598.085058     307.997702   

       osrm_distance  segment_actual_time  segment_osrm_time  \
count  144846.000000        144846.000000      144846.000000   
mean      284.750969            36.207427          18.507304   
min         9.008200             0.000000           0.000000   
25%        29.909925            20.000000          11.000000   
50%        78.524600            29.000000          17.000000   
75%       343.062075            40.000000          22.000000   
max      2326.199100          3051.000000        1611.000000   
std       421.101831            53.561259          14.775870   

       segment_osrm_distance  
count          144846.000000  
mean               22.828528  
min                 0.000000  
25%                12.070100  
50%                23.513000  
75%                27.812975  
max              2191.403700  
std                17.860268

In [344]:
df.isna().sum()

data                                0
trip_creation_time                  0
route_schedule_uuid                 0
route_type                          0
trip_uuid                           0
source_center                       0
source_name                       293
destination_center                  0
destination_name                  261
od_start_time                       0
od_end_time                         0
start_scan_to_end_scan              0
actual_distance_to_destination      0
actual_time                         0
osrm_time                           0
osrm_distance                       0
segment_actual_time                 0
segment_osrm_time                   0
segment_osrm_distance               0
dtype: int64

Now negative values of negative segment time are removed.

#### Merging rows and aggregaion of data

In [345]:
df.head(5)

data         trip_creation_time  \
0  training 2018-09-20 02:35:36.476840   
1  training 2018-09-20 02:35:36.476840   
2  training 2018-09-20 02:35:36.476840   
3  training 2018-09-20 02:35:36.476840   
4  training 2018-09-20 02:35:36.476840   

                                 route_schedule_uuid route_type  \
0  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
1  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
2  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
3  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
4  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   

                 trip_uuid source_center                 source_name  \
0  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
1  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
2  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
3  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
4  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   

  destination_center               destination_name  \
0       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
1       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
2       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
3       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
4       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   

               od_start_time                od_end_time  \
0 2018-09-20 03:21:32.418600 2018-09-20 04:47:45.236797   
1 2018-09-20 03:21:32.418600 2018-09-20 04:47:45.236797   
2 2018-09-20 03:21:32.418600 2018-09-20 04:47:45.236797   
3 2018-09-20 03:21:32.418600 2018-09-20 04:47:45.236797   
4 2018-09-20 03:21:32.418600 2018-09-20 04:47:45.236797   

   start_scan_to_end_scan  actual_distance_to_destination  actual_time  \
0                    86.0                       10.435660         14.0   
1                    86.0                       18.936842         24.0   
2                    86.0                       27.637279         40.0   
3                    86.0                       36.118028         62.0   
4                    86.0                       39.386040         68.0   

   osrm_time  osrm_distance  segment_actual_time  segment_osrm_time  \
0       11.0        11.9653                 14.0               11.0   
1       20.0        21.7243                 10.0                9.0   
2       28.0        32.5395                 16.0                7.0   
3       40.0        45.5620                 21.0               12.0   
4       44.0        54.2181                  6.0                5.0   

   segment_osrm_distance  
0                11.9653  
1                 9.7590  
2                10.8152  
3                13.0224  
4                 3.9153

##### Merging row based on trip_uuid, source center and destination center as df1

In [346]:
# name of columns for aggregation
group_by_columns = ['trip_uuid', 'source_center', 'destination_center']

df1 = df.groupby(by = group_by_columns, as_index= False ).agg(
                                                                  {
                                                                      'data' : 'first',
                                                                      'trip_creation_time' : 'first',
                                                                      'route_type': 'first',
                                                                      'source_name' : 'first',
                                                                      'destination_name' : 'last',
                                                                      'od_start_time' : 'first',
                                                                      'od_end_time' : 'first',
                                                                      'start_scan_to_end_scan' : 'first',
                                                                      'actual_distance_to_destination' : 'last', #cumulative distance so using the last for taking end value
                                                                      'actual_time' : 'last', #cumulative time for trip till that point so using last for end value
                                                                      'osrm_time' : 'last', #cumualative time
                                                                      'osrm_distance' : 'last', #cumulative distance so taking last
                                                                      'segment_actual_time' : 'sum', #segment time so doing summation
                                                                      'segment_osrm_time' : 'sum', #segment time so doing the sum for whole trip
                                                                      'segment_osrm_distance' : 'sum' #segment distance so doing the summation


                                                                  }
                                                              )
df1.head()

trip_uuid source_center destination_center      data  \
0  trip-153671041653548748  IND209304AAA       IND000000ACB  training   
1  trip-153671041653548748  IND462022AAA       IND209304AAA  training   
2  trip-153671042288605164  IND561203AAB       IND562101AAA  training   
3  trip-153671042288605164  IND572101AAA       IND561203AAB  training   
4  trip-153671043369099517  IND000000ACB       IND160002AAC  training   

          trip_creation_time route_type                         source_name  \
0 2018-09-12 00:00:16.535741        FTL  Kanpur_Central_H_6 (Uttar Pradesh)   
1 2018-09-12 00:00:16.535741        FTL  Bhopal_Trnsport_H (Madhya Pradesh)   
2 2018-09-12 00:00:22.886430    Carting   Doddablpur_ChikaDPP_D (Karnataka)   
3 2018-09-12 00:00:22.886430    Carting       Tumkur_Veersagr_I (Karnataka)   
4 2018-09-12 00:00:33.691250        FTL       Gurgaon_Bilaspur_HB (Haryana)   

                     destination_name              od_start_time  \
0       Gurgaon_Bilaspur_HB (Haryana) 2018-09-12 16:39:46.858469   
1  Kanpur_Central_H_6 (Uttar Pradesh) 2018-09-12 00:00:16.535741   
2   Chikblapur_ShntiSgr_D (Karnataka) 2018-09-12 02:03:09.655591   
3   Doddablpur_ChikaDPP_D (Karnataka) 2018-09-12 00:00:22.886430   
4      Chandigarh_Mehmdpur_H (Punjab) 2018-09-14 03:40:17.106733   

                 od_end_time  start_scan_to_end_scan  \
0 2018-09-13 13:40:23.123744                  1260.0   
1 2018-09-12 16:39:46.858469                   999.0   
2 2018-09-12 03:01:59.598855                    58.0   
3 2018-09-12 02:03:09.655591                   122.0   
4 2018-09-14 17:34:55.442454                   834.0   

   actual_distance_to_destination  actual_time  osrm_time  osrm_distance  \
0                      383.759164        732.0      329.0       446.5496   
1                      440.973689        830.0      388.0       544.8027   
2                       24.644021         47.0       26.0        28.1994   
3                       48.542890         96.0       42.0        56.9116   
4                      237.439610        611.0      212.0       281.2109   

   segment_actual_time  segment_osrm_time  segment_osrm_distance  
0                728.0              534.0               670.6205  
1                820.0              474.0               649.8528  
2                 46.0               26.0                28.1995  
3                 95.0               39.0                55.9899  
4                608.0              231.0               317.7408

#### Adding the od_total_time

Calculate the time taken between od_start_time and od_end_time and keep it as a feature. Drop the original columns, if required

In [347]:
#calculating the total trip time
df1['od_total_time'] = df1['od_end_time'] - df1['od_start_time']

#dropping the original columns from the dataframe
df1.drop(['od_start_time', 'od_end_time'], inplace = True, axis = 1)

#convert the total time into seconds
df1['od_total_time'] = ((df1['od_total_time'].dt.total_seconds())/ 60).round(2)

df1.head(2) #check the value in new column for total trip time

trip_uuid source_center destination_center      data  \
0  trip-153671041653548748  IND209304AAA       IND000000ACB  training   
1  trip-153671041653548748  IND462022AAA       IND209304AAA  training   

          trip_creation_time route_type                         source_name  \
0 2018-09-12 00:00:16.535741        FTL  Kanpur_Central_H_6 (Uttar Pradesh)   
1 2018-09-12 00:00:16.535741        FTL  Bhopal_Trnsport_H (Madhya Pradesh)   

                     destination_name  start_scan_to_end_scan  \
0       Gurgaon_Bilaspur_HB (Haryana)                  1260.0   
1  Kanpur_Central_H_6 (Uttar Pradesh)                   999.0   

   actual_distance_to_destination  actual_time  osrm_time  osrm_distance  \
0                      383.759164        732.0      329.0       446.5496   
1                      440.973689        830.0      388.0       544.8027   

   segment_actual_time  segment_osrm_time  segment_osrm_distance  \
0                728.0              534.0               670.6205   
1                820.0              474.0               649.8528   

   od_total_time  
0        1260.60  
1         999.51

##### Merging row based on trip_uuid as df2

In [348]:
df2 = df1.groupby(by = 'trip_uuid', as_index = False).agg({'source_center' : 'first',
                                                           'destination_center' : 'last',
                                                           'data' : 'first',
                                                           'route_type' : 'first',
                                                           'trip_creation_time' : 'first',
                                                           'source_name' : 'first',
                                                           'destination_name' : 'last',
                                                           'od_total_time' : 'sum',
                                                           'start_scan_to_end_scan' : 'sum',
                                                           'actual_distance_to_destination' : 'sum',
                                                           'actual_time' : 'sum',
                                                           'osrm_time' : 'sum',
                                                           'osrm_distance' : 'sum',
                                                           'segment_actual_time' : 'sum',
                                                           'segment_osrm_time' : 'sum',
                                                           'segment_osrm_distance' : 'sum'})
df2.head()

trip_uuid source_center destination_center      data  \
0  trip-153671041653548748  IND209304AAA       IND209304AAA  training   
1  trip-153671042288605164  IND561203AAB       IND561203AAB  training   
2  trip-153671043369099517  IND000000ACB       IND000000ACB  training   
3  trip-153671046011330457  IND400072AAB       IND401104AAA  training   
4  trip-153671052974046625  IND583101AAA       IND583119AAA  training   

  route_type         trip_creation_time                         source_name  \
0        FTL 2018-09-12 00:00:16.535741  Kanpur_Central_H_6 (Uttar Pradesh)   
1    Carting 2018-09-12 00:00:22.886430   Doddablpur_ChikaDPP_D (Karnataka)   
2        FTL 2018-09-12 00:00:33.691250       Gurgaon_Bilaspur_HB (Haryana)   
3    Carting 2018-09-12 00:01:00.113710            Mumbai Hub (Maharashtra)   
4        FTL 2018-09-12 00:02:09.740725              Bellary_Dc (Karnataka)   

                     destination_name  od_total_time  start_scan_to_end_scan  \
0  Kanpur_Central_H_6 (Uttar Pradesh)        2260.11                  2259.0   
1   Doddablpur_ChikaDPP_D (Karnataka)         181.61                   180.0   
2       Gurgaon_Bilaspur_HB (Haryana)        3934.36                  3933.0   
3      Mumbai_MiraRd_IP (Maharashtra)         100.49                   100.0   
4       Sandur_WrdN1DPP_D (Karnataka)         718.34                   717.0   

   actual_distance_to_destination  actual_time  osrm_time  osrm_distance  \
0                      824.732854       1562.0      717.0       991.3523   
1                       73.186911        143.0       68.0        85.1110   
2                     1927.404273       3347.0     1740.0      2354.0665   
3                       17.175274         59.0       15.0        19.6800   
4                      127.448500        341.0      117.0       146.7918   

   segment_actual_time  segment_osrm_time  segment_osrm_distance  
0               1548.0             1008.0              1320.4733  
1                141.0               65.0                84.1894  
2               3308.0             1941.0              2545.2678  
3                 59.0               16.0                19.8766  
4                340.0              115.0               146.7919

### Feature Generation

#### Extract state, City and Place from Source & Destination

In [349]:
("Kanpur_Central_H_6 (Uttar Pradesh)").split('_')

['Kanpur', 'Central', 'H', '6 (Uttar Pradesh)']

In [350]:
("Kanpur_Central_H_6 (Uttar Pradesh)").split('_')[0] #city

'Kanpur'

In [351]:
("Kanpur_Central_H_6 (Uttar Pradesh)").split('_')[1] #place

'Central'

In [352]:
("Kanpur_Central_H_6 (Uttar Pradesh)").split('(')[1]

'Uttar Pradesh)'

In [353]:
("Kanpur_Central_H_6 (Uttar Pradesh)").split('(')[1][:-1]  #state

'Uttar Pradesh'

In [354]:
#we will be using the df2 dataframe that is been grouped at trip_uuid level

# Extracting state, city and place name from source_name column value e.g. Kanpur_Central_H_6 (Uttar Pradesh)

# Source Name: Split and extract features out of destination. City-place-code (State)
df2['source_state'] = df2['source_name'].apply(lambda x : str(x).split('(')[1][:-1] if '(' in str(x) else None) #state
df2['source_city'] = df2['source_name'].apply(lambda x : str(x).split('_')[0]) #city
#df2['source_place'] = df2['source_name'].apply(lambda x : str(x).split('_')[1]) #place

# Destination Name: Split and extract features out of destination. City-place-code (State)
df2['destination_state'] = df2['destination_name'].apply(lambda x : str(x).split('(')[1][:-1] if '(' in str(x) else None) #state
df2['destination_city'] = df2['destination_name'].apply(lambda x : str(x).split('_')[0]) #city
#df2['destination_place'] = df2['destination_name'].apply(lambda x : str(x).split('_')[1]) #place

df2.head()

trip_uuid source_center destination_center      data  \
0  trip-153671041653548748  IND209304AAA       IND209304AAA  training   
1  trip-153671042288605164  IND561203AAB       IND561203AAB  training   
2  trip-153671043369099517  IND000000ACB       IND000000ACB  training   
3  trip-153671046011330457  IND400072AAB       IND401104AAA  training   
4  trip-153671052974046625  IND583101AAA       IND583119AAA  training   

  route_type         trip_creation_time                         source_name  \
0        FTL 2018-09-12 00:00:16.535741  Kanpur_Central_H_6 (Uttar Pradesh)   
1    Carting 2018-09-12 00:00:22.886430   Doddablpur_ChikaDPP_D (Karnataka)   
2        FTL 2018-09-12 00:00:33.691250       Gurgaon_Bilaspur_HB (Haryana)   
3    Carting 2018-09-12 00:01:00.113710            Mumbai Hub (Maharashtra)   
4        FTL 2018-09-12 00:02:09.740725              Bellary_Dc (Karnataka)   

                     destination_name  od_total_time  start_scan_to_end_scan  \
0  Kanpur_Central_H_6 (Uttar Pradesh)        2260.11                  2259.0   
1   Doddablpur_ChikaDPP_D (Karnataka)         181.61                   180.0   
2       Gurgaon_Bilaspur_HB (Haryana)        3934.36                  3933.0   
3      Mumbai_MiraRd_IP (Maharashtra)         100.49                   100.0   
4       Sandur_WrdN1DPP_D (Karnataka)         718.34                   717.0   

   ...  actual_time  osrm_time  osrm_distance  segment_actual_time  \
0  ...       1562.0      717.0       991.3523               1548.0   
1  ...        143.0       68.0        85.1110                141.0   
2  ...       3347.0     1740.0      2354.0665               3308.0   
3  ...         59.0       15.0        19.6800                 59.0   
4  ...        341.0      117.0       146.7918                340.0   

   segment_osrm_time  segment_osrm_distance   source_state  \
0             1008.0              1320.4733  Uttar Pradesh   
1               65.0                84.1894      Karnataka   
2             1941.0              2545.2678        Haryana   
3               16.0                19.8766    Maharashtra   
4              115.0               146.7919      Karnataka   

                source_city destination_state destination_city  
0                    Kanpur     Uttar Pradesh           Kanpur  
1                Doddablpur         Karnataka       Doddablpur  
2                   Gurgaon           Haryana          Gurgaon  
3  Mumbai Hub (Maharashtra)       Maharashtra           Mumbai  
4                   Bellary         Karnataka           Sandur  

[5 rows x 21 columns]

In [355]:
df2['source_state'].unique()

array(['Uttar Pradesh', 'Karnataka', 'Haryana', 'Maharashtra',
       'Tamil Nadu', 'Gujarat', 'Delhi', 'Telangana', 'Rajasthan',
       'Assam', 'Madhya Pradesh', 'West Bengal', 'Andhra Pradesh',
       'Punjab', 'Chandigarh', 'Goa', 'Jharkhand', 'Pondicherry',
       'Orissa', 'Uttarakhand', 'Himachal Pradesh', 'Kerala',
       'Arunachal Pradesh', 'Bihar', 'Chhattisgarh',
       'Dadra and Nagar Haveli', 'Jammu & Kashmir', 'Mizoram', 'Nagaland',
       None], dtype=object)

In [356]:
df2['destination_state'].unique()

array(['Uttar Pradesh', 'Karnataka', 'Haryana', 'Maharashtra',
       'Tamil Nadu', 'Gujarat', 'Delhi', 'Telangana', 'Rajasthan',
       'Madhya Pradesh', 'Assam', 'West Bengal', 'Andhra Pradesh',
       'Punjab', 'Chandigarh', 'Dadra and Nagar Haveli', 'Orissa',
       'Bihar', 'Jharkhand', 'Goa', 'Uttarakhand', 'Himachal Pradesh',
       'Kerala', 'Arunachal Pradesh', 'Mizoram', 'Chhattisgarh',
       'Jammu & Kashmir', 'Nagaland', 'Meghalaya', 'Tripura', None,
       'Daman & Diu'], dtype=object)

State names for source and destination looks good

In [357]:
df2['source_city'].unique()[:50]

array(['Kanpur', 'Doddablpur', 'Gurgaon', 'Mumbai Hub (Maharashtra)',
       'Bellary', 'Chennai', 'HBR Layout PC (Karnataka)', 'Surat',
       'Delhi', 'Pune', 'FBD', 'Shirala', 'Hyderabad', 'Thirumalagiri',
       'Gulbarga', 'Jaipur', 'Allahabad', 'Guwahati', 'Narsinghpur',
       'Shrirampur', 'Hoogly', 'Madakasira', 'Sonari', 'Bengaluru',
       'Dindigul', 'Jalandhar', 'Faridabad', 'Chandigarh', 'Deoli',
       'Pandharpur', 'CCU', 'Bhandara', 'Kurnool', 'Bhiwandi', 'Bhatinda',
       'RoopNagar', 'Bantwal', 'Lalru', 'Kadi', 'Shahdol', 'Gangakher',
       'Durgapur', 'Vapi', 'Jamjodhpur', 'Jetpur', 'Mehsana', 'Jabalpur',
       'Junagadh', 'Gundlupet', 'Mysore'], dtype=object)

In [358]:
df2['destination_city'].unique()[:50]

array(['Kanpur', 'Doddablpur', 'Gurgaon', 'Mumbai', 'Sandur', 'Chennai',
       'HBR Layout PC (Karnataka)', 'Surat', 'Delhi',
       'PNQ Rahatani DPC (Maharashtra)', 'Faridabad (Haryana)',
       'Ratnagiri', 'Bangalore', 'Hyderabad', 'Aland', 'Jaipur', 'Satna',
       'Janakpuri (Delhi)', 'Guwahati', 'Bareli', 'Nashik', 'Hooghly',
       'Puttaprthi', 'Sivasagar', 'Bengaluru', 'Palani', 'Jalandhar',
       'Chandigarh', 'Yavatmal', 'Sangola', 'Kolkata', 'Savner',
       'Kurnool', 'FBD', 'Bhatinda', 'Bhiwandi', 'Barnala', 'Murbad',
       'Kadaba', 'Gulbarga', 'Naraingarh', 'Ludhiana', 'Kadi', 'Jabalpur',
       'MAA', 'Gangakher', 'Bankura', 'Silvassa', 'Porbandar', 'Jetpur'],
      dtype=object)

With the city name there is some problem, at some places name of state is added with it.

In [359]:
df2['source_city'] = df2['source_city'].apply(lambda x : str(x).split()[0])
df2['destination_city'] = df2['destination_city'].apply(lambda x : str(x).split()[0])
df2.sample(10)

trip_uuid source_center destination_center      data  \
587    trip-153678476032423204  IND562132AAA       IND560300AAA  training   
1747   trip-153690681610502349  IND110064AAA       IND000000ACB  training   
7710   trip-153763761783703072  IND382430AAB       IND421302AAG  training   
2330   trip-153697823038467653  IND387001AAA       IND387001AAA  training   
5066   trip-153731180533627844  IND000000ACB       IND131028AAB  training   
8548   trip-153774263614858836  IND462022AAA       IND462022AAA  training   
3412   trip-153712554195755989  IND421302AAG       IND562132AAA  training   
3817   trip-153716795842522836  IND110014AAA       IND110024AAA  training   
11908  trip-153817885327966424  IND121004AAB       IND121004AAA      test   
4997   trip-153730763549066095  IND273014AAB       IND277201AAA  training   

      route_type         trip_creation_time  \
587      Carting 2018-09-12 20:39:20.324489   
1747     Carting 2018-09-14 06:33:36.105259   
7710         FTL 2018-09-22 17:33:37.837275   
2330     Carting 2018-09-15 02:23:50.385028   
5066     Carting 2018-09-18 23:03:25.336546   
8548         FTL 2018-09-23 22:43:56.148841   
3412         FTL 2018-09-16 19:19:01.957793   
3817     Carting 2018-09-17 07:05:58.425509   
11908    Carting 2018-09-28 23:54:13.280015   
4997         FTL 2018-09-18 21:53:55.490956   

                                 source_name  \
587         Bangalore_Nelmngla_H (Karnataka)   
1747               Delhi_Mayapuri_PC (Delhi)   
7710            Ahmedabad_East_H_1 (Gujarat)   
2330                     Nadiad_DC (Gujarat)   
5066           Gurgaon_Bilaspur_HB (Haryana)   
8548      Bhopal_Trnsport_H (Madhya Pradesh)   
3412       Bhiwandi_Mankoli_HB (Maharashtra)   
3817                    Delhi_Bhogal (Delhi)   
11908           FBD_Balabhgarh_DPC (Haryana)   
4997   Gorakhpur_Matriprm_IP (Uttar Pradesh)   

                         destination_name  od_total_time  \
587     Bengaluru_KGAirprt_HB (Karnataka)         314.65   
1747        Gurgaon_Bilaspur_HB (Haryana)         720.29   
7710    Bhiwandi_Mankoli_HB (Maharashtra)        1537.75   
2330                  Nadiad_DC (Gujarat)         201.43   
5066           Sonipat_Kundli_H (Haryana)         317.44   
8548   Bhopal_Trnsport_H (Madhya Pradesh)        2872.67   
3412     Bangalore_Nelmngla_H (Karnataka)        1427.13   
3817              Delhi_Lajpat_IP (Delhi)          57.72   
11908      Faridabad_Blbgarh_DC (Haryana)          76.67   
4997    Bariya_BgnprDPP_D (Uttar Pradesh)         311.55   

       start_scan_to_end_scan  ...  actual_time  osrm_time  osrm_distance  \
587                     314.0  ...         55.0       46.0        36.8545   
1747                    720.0  ...        693.0       53.0        57.2795   
7710                   1536.0  ...        960.0      405.0       541.7943   
2330                    200.0  ...        128.0       64.0        74.5832   
5066                    317.0  ...        158.0       98.0       122.9206   
8548                   2872.0  ...       2510.0     1198.0      1680.0168   
3412                   1427.0  ...       1192.0      712.0       970.9006   
3817                     57.0  ...         24.0        7.0         9.0729   
11908                    76.0  ...         20.0       12.0        16.1132   
4997                    311.0  ...        245.0      122.0       153.4156   

       segment_actual_time  segment_osrm_time  segment_osrm_distance  \
587                   54.0               65.0                65.8588   
1747                 691.0               50.0                57.2796   
7710                 955.0              412.0               556.2722   
2330                 125.0               64.0                74.5831   
5066                 155.0              191.0               129.1677   
8548                2490.0             1534.0              1837.2194   
3412                1174.0              762.0               982.8149   
3817                  24.0               

Now the city data is also clean. So we are good for some EDA analysis.

#### Extract Date, year, month, day, hour and week from trip creation time

In [360]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14817 entries, 0 to 14816
Data columns (total 21 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   trip_uuid                       14817 non-null  object        
 1   source_center                   14817 non-null  object        
 2   destination_center              14817 non-null  object        
 3   data                            14817 non-null  category      
 4   route_type                      14817 non-null  category      
 5   trip_creation_time              14817 non-null  datetime64[ns]
 6   source_name                     14807 non-null  object        
 7   destination_name                14809 non-null  object        
 8   od_total_time                   14817 non-null  float64       
 9   start_scan_to_end_scan          14817 non-null  float64       
 10  actual_distance_to_destination  14817 non-null  float64       
 11  ac

In [361]:
# since 'trip_creation_time' is already a datetime column so we don't need to use the pd.to_datetime here again.

df2['trip_creation_date'] = df2['trip_creation_time'].dt.date
df2['trip_creation_year'] = df2['trip_creation_time'].dt.year.astype('int16')
df2['trip_creation_month'] = df2['trip_creation_time'].dt.month.astype('int8')
df2['trip_creation_day'] = df2['trip_creation_time'].dt.day.astype('int8')
df2['trip_creation_week'] = df2['trip_creation_time'].dt.isocalendar().week.astype('int8')
df2['trip_creation_hour'] = df2['trip_creation_time'].dt.hour.astype('int8')

df2.sample(5)

trip_uuid source_center destination_center      data  \
5585   trip-153738482530900806  IND562132AAA       IND560099AAB  training   
8632   trip-153774865685588881  IND614620AAA       IND623407AAA  training   
8139   trip-153767748006206022  IND712311AAA       IND700065AAA  training   
11937  trip-153818067015264507  IND282001AAF       IND282002AAD      test   
4816   trip-153729413047777264  IND834002AAB       IND832109AAB  training   

      route_type         trip_creation_time  \
5585     Carting 2018-09-19 19:20:25.309225   
8632     Carting 2018-09-24 00:24:16.856131   
8139     Carting 2018-09-23 04:38:00.062339   
11937        FTL 2018-09-29 00:24:30.152895   
4816         FTL 2018-09-18 18:08:50.478028   

                               source_name  \
5585      Bangalore_Nelmngla_H (Karnataka)   
8632   Manamelkudi_TmpleSrt_D (Tamil Nadu)   
8139      Kolkata_Dankuni_HB (West Bengal)   
11937     Agra_Central_D_3 (Uttar Pradesh)   
4816                Ranchi_Hub (Jharkhand)   

                           destination_name  od_total_time  \
5585      Bengaluru_Bomsndra_HB (Karnataka)         434.39   
8632   Thiruvadanai_RamnadRD_D (Tamil Nadu)         353.96   
8139       CCU_Beliaghata_DPC (West Bengal)         110.60   
11937          Dholpur_GtRoad_D (Rajasthan)         422.66   
4816     Jamshedpur_Central_I_3 (Jharkhand)         328.04   

       start_scan_to_end_scan  ...   source_state  source_city  \
5585                    434.0  ...      Karnataka    Bangalore   
8632                    353.0  ...     Tamil Nadu  Manamelkudi   
8139                    110.0  ...    West Bengal      Kolkata   
11937                   420.0  ...  Uttar Pradesh         Agra   
4816                    328.0  ...      Jharkhand       Ranchi   

       destination_state  destination_city  trip_creation_date  \
5585           Karnataka         Bengaluru          2018-09-19   
8632          Tamil Nadu      Thiruvadanai          2018-09-24   
8139         West Bengal               CCU          2018-09-23   
11937          Rajasthan           Dholpur          2018-09-29   
4816           Jharkhand        Jamshedpur          2018-09-18   

       trip_creation_year  trip_creation_month trip_creation_day  \
5585                 2018                    9                19   
8632                 2018                    9                24   
8139                 2018                    9                23   
11937                2018                    9                29   
4816                 2018                    9                18   

      trip_creation_week trip_creation_hour  
5585                  38                 19  
8632                  39                  0  
8139                  38                  4  
11937                 39                  0  
4816                  38                 18  

[5 rows x 27 columns]

In [362]:
df2.describe().T

count                           mean  \
trip_creation_time                14817  2018-09-22 12:44:19.555167744   
od_total_time                   14817.0                      531.69763   
start_scan_to_end_scan          14817.0                     530.810016   
actual_distance_to_destination  14817.0                     164.477838   
actual_time                     14817.0                     357.143754   
osrm_time                       14817.0                     161.384018   
osrm_distance                   14817.0                     204.344689   
segment_actual_time             14817.0                      353.95161   
segment_osrm_time               14817.0                     180.921172   
segment_osrm_distance           14817.0                        223.164   
trip_creation_year              14817.0                         2018.0   
trip_creation_month             14817.0                       9.120672   
trip_creation_day               14817.0                       18.37079   
trip_creation_week              14817.0                      38.295944   
trip_creation_hour              14817.0                      12.449821   

                                                       min  \
trip_creation_time              2018-09-12 00:00:16.535741   
od_total_time                                        23.46   
start_scan_to_end_scan                                23.0   
actual_distance_to_destination                    9.002461   
actual_time                                            9.0   
osrm_time                                              6.0   
osrm_distance                                       9.0729   
segment_actual_time                                    9.0   
segment_osrm_time                                      6.0   
segment_osrm_distance                               9.0729   
trip_creation_year                                  2018.0   
trip_creation_month                                    9.0   
trip_creation_day                                      1.0   
trip_creation_week                                    37.0   
trip_creation_hour                                     0.0   

                                                          25%  \
trip_creation_time              2018-09-17 02:51:25.129125888   
od_total_time                                          149.93   
start_scan_to_end_scan                                  149.0   
actual_distance_to_destination                      22.837239   
actual_time                                              67.0   
osrm_time                                                29.0   
osrm_distance                                         30.8192   
segment_actual_time                                      66.0   
segment_osrm_time                                        31.0   
segment_osrm_distance                                 32.6545   
trip_creation_year                                     2018.0   
trip_creation_month                                       9.0   
trip_creation_day                                        14.0   
trip_creation_week                                       38.0   
trip_creation_hour                                        4.0   

                                                          50%  \
trip_creation_time              2018-09-22 04:02:35.066945024   
od_total_time                                          280.77   
start_scan_to_end_scan                                  280.0   
actual_distance_to_destination                      48.474072   
actual_time                                             149.0   
osrm_time                                                60.0   
osrm_distance                                         65.6188   
segment_actual_time                                     147.0   
segment_osrm_time                                        65.0   
segment_osrm_distance                                 70.1544   
trip_creation_year                                     2018.0   
trip_creation_month                

In [363]:
df2.describe(include = 'object').T

count unique                            top  freq
trip_uuid           14817  14817        trip-153671041653548748     1
source_center       14817    938                   IND000000ACB  1063
destination_center  14817   1042                   IND000000ACB   821
source_name         14807    933  Gurgaon_Bilaspur_HB (Haryana)  1063
destination_name    14809   1034  Gurgaon_Bilaspur_HB (Haryana)   821
source_state        14807     29                    Maharashtra  2714
source_city         14817    717                        Gurgaon  1139
destination_state   14809     31                    Maharashtra  2561
destination_city    14817    841                         Mumbai  1202
trip_creation_date  14817     22                     2018-09-18   791

### Univariate & Bi-variate Analysis

### Hypothesis testing

### Outlier Detection

### Outlier Treatment

### Encoding

### Normalisation/ Standardization

## Business Insights

### Recommendations